src/abi_x86_64.cpp

```diff
@@ -239,7 +240,9 @@ Type *preferred_llvm_type(jl_datatype_t *dt, bool isret, LLVMContext &ctx) const
                 types[0] = Type::getIntNTy(ctx, nbits);
             break;
         case Sse:
-            if (size <= 4)
+            if (size <= 2)
+                types[0] = Type::getHalfTy(ctx);
```

Comment:
It's passed in the same way, yes.

This should also fix how Float16 is passed. I guess we weren't interfacing with C libraries passing `half` yet.

Reply:
So that's the ABI break part 3. Because we currently assume it gets passed as an uint.

Reply:
> So that's the ABI break part 3.

I must be missing something; what are ABI breaks part 1 and 2? I already addressed Valentin's comments on the sign/zero extension.

Reply:
Float16 used to be passed as an uint16. In GCC 12 it changed to be passed as a float. When that happened we had to add some annoying things. Then LLVM 15 did the change and we had to finish adapting. This change would then change the ccall ABI which I think is the last remaining thing using the uint16 ABI. (The conversion functions we implement actually still do and we have to do some hacks in LLVM because we don't want to bump the minimum compiler to GCC 12. Though I guess we could fake the ABI and use float)

Reply:
> And I think the conversion functions still do that, because otherwise building julia would require very new compilers.

Ah yes, I forgot to update the Float16 conversion functions. We should be able to mimic the new ABI using the `float` trick I did with the BFloat16 conversion functions though, so maybe we should just switch to the new ABI?

Reply:
That would be nice. We also remove some of the hacks I had to put in place.

Reply:
Though I'm not sure if the ABI change is true for every architecture.

Reply:
I reverted the ABI change (not in this exact location, but by not setting the Sse flag for Float16 above).

I'll investigate simplifying the Float16 ABI, but that proves to be a more tricky change, so I'd prefer to get the BFloat16 support merged first.